### PR TITLE
Features/more game state

### PR DESCRIPTION
**Pull request by DavidBetteridge**

### CODE REVIEW

File: src/NPacMan.Game/Actions.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
+﻿using Automatonymous;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace NPacMan.Game
+{
+    internal static class Actions
```
Comment (DavidBetteridge) [inline src/NPacMan.Game/Actions.cs:9]: This class should be the only place where game state can be changed???

File: src/NPacMan.Game/GameStateMachine.cs [hunk 2]
```diff
@@ -46,40 +45,46 @@ public GameStateMachine(IGameActions game, IGameSettings settings, GameNotificat
...
                                 .Then(context => context.Instance.TimeToChangeState = context.Instance.LastTick.AddSeconds(4))
                                 .Then(context => gameNotifications.Publish(GameNotification.Dying)));
 
             During(Dying,
                 When(Tick, context => context.Data.Now >= context.Instance.TimeToChangeState)
                     .IfElse(context => context.Instance.Lives > 0,
                         binder => binder.TransitionTo(Respawning),
-                        binder => binder.TransitionTo(Dead)));
+                        binder => binder.TransitionTo(Dead)),
+                When(CoinCollision)
```
Comment (DavidBetteridge) [inline src/NPacMan.Game/GameStateMachine.cs:82]: This isn't great

File: src/NPacMan.Game/GameStateMachine.cs [hunk 3]
```diff
@@ -46,40 +45,46 @@ public GameStateMachine(IGameActions game, IGameSettings settings, GameNotificat
...
-                    .Then(x => game.MakeGhostsNotEdible())
+                    .Then(context => Actions.MakeGhostsNotEdible(context.Instance))
                     .TransitionTo(Scatter));
 
             During(Scatter, GhostChase, Frightened,
                 When(Tick)
-                    .ThenAsync(async context => await game.MoveGhosts(context, this))
-                    .Then(context => game.MovePacMan(context, this)),
+                    .ThenAsync(async context => await Actions.MoveGhosts(game, context.Instance, context, this))
+                    .ThenAsync(async context => await Actions.MovePacMan(settings, context.Instance, context, this)),
```
Comment (DavidBetteridge) [inline src/NPacMan.Game/GameStateMachine.cs:54]: This should be async

File: src/NPacMan.Game/Actions.cs (hunk 1 above)
Comment (kevbite) [reply src/NPacMan.Game/Actions.cs:9]: Yeah, That'd be nice 👍

File: src/NPacMan.Game/Actions.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            ApplyToGhosts(gameState, ghost => ghost.Move(game));
+
+            var ghosts = GhostsCollidedWithPacMan(gameState);
+            foreach (var ghost in ghosts)
+            {
+                await context.Raise(gameStateMachine.GhostCollision, new GhostCollision(ghost));
+            }
+        }
+
+        public static void RemoveCoin(GameState gameState, CellLocation location)
```
Comment (kevbite) [inline src/NPacMan.Game/Actions.cs:122]: I feel this method should just go on GameState?

File: src/NPacMan.Game/Actions.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            // Note - this is not the same as gameState.RemainingCoins = gameState.RemainingCoins.Remove(location)
+            // We have to allow for the UI to be iterating over the list whilst we are removing elements from it.
+            gameState.RemainingCoins = gameState.RemainingCoins.Where(c => c != location).ToList();
+        }
+
+        public static void RemovePowerPill(GameState gameState, CellLocation location)
```
Comment (kevbite) [inline src/NPacMan.Game/Actions.cs:129]: Same as above be better if we moved this to game state.

File: src/NPacMan.Game/Actions.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            gameState.GhostsVisible = false;
+        }
+
+        private static IEnumerable<Ghost> GhostsCollidedWithPacMan(GameState gameState)
+        {
+            return gameState.Ghosts.Values.Where(ghost => ghost.Location == gameState.PacMan.Location);
+        }
+
+        private static void ApplyToGhosts(GameState gameState, Func<Ghost, Ghost> action)
```
Comment (kevbite) [inline src/NPacMan.Game/Actions.cs:26]: Should this be inside the game state too?

File: src/NPacMan.Game/GameStateMachine.cs (hunk 2 above)
Comment (kevbite) [reply src/NPacMan.Game/GameStateMachine.cs:82]:
```chsarp
When(Dying,
    Ignore(CoinCollision));
```

Comment (DavidBetteridge) [reply src/NPacMan.Game/GameStateMachine.cs]: Much better

**Pull request merged**